File: bindings/php/src/Signature.php
<?php
/**
 * libpqc-dyber - Post-Quantum Cryptography Library
 * Copyright (c) 2024-2026 Dyber, Inc.
 * SPDX-License-Identifier: Apache-2.0 OR MIT
 *
 * Signature class.
 */

declare(strict_types=1);

namespace Dyber\PQC;

use RuntimeException;

final class Signature
{
    /** @var \FFI\CData */
    private $handle;
    private string $algorithm;
    private bool $freed = false;

    /**
     * Create a Signature context for the specified algorithm.
     *
     * @param string $algorithm Algorithm name, e.g. "ML-DSA-65"
     * @throws RuntimeException if the algorithm is not supported
     */
    public function __construct(string $algorithm)
    {
        $this->algorithm = $algorithm;
        $ffi = PQC::ffi();
        $this->handle = $ffi->pqc_sig_new($algorithm);
        if (\FFI::isNull($this->handle)) {
            throw new RuntimeException("Unsupported signature algorithm: {$algorithm}");
        }
    }

    public function __destruct()
    {
        if (!$this->freed) {
            PQC::ffi()->pqc_sig_free($this->handle);
        }
    }

    public function free(): void
    {
        if (!$this->freed) {
            PQC::ffi()->pqc_sig_free($this->handle);
            $this->freed = true;
        }
    }

    public function algorithm(): string
    {
        return $this->algorithm;
    }

    public function publicKeySize(): int
    {
        return PQC::ffi()->pqc_sig_public_key_size($this->handle);
    }

    public function secretKeySize(): int
    {
        return PQC::ffi()->pqc_sig_secret_key_size($this->handle);
    }

    public function maxSignatureSize(): int
    {
        return PQC::ffi()->pqc_sig_max_signature_size($this->handle);
    }

    /**
     * Generate a keypair.
     *
     * @return array{publicKey: string, secretKey: string}
     */
    public function keygen(): array
    {
        $ffi = PQC::ffi();
        $pkSize = $this->publicKeySize();
        $skSize = $this->secretKeySize();

        $pk = $ffi->new("uint8_t[{$pkSize}]");
        $sk = $ffi->new("uint8_t[{$skSize}]");

        $rc = $ffi->pqc_sig_keygen($this->handle, $pk, $sk);
        if ($rc !== 0) {
            throw new RuntimeException("Signature keygen failed: " . $ffi->pqc_status_string($rc));
        }

        return [
            'publicKey' => \FFI::string($pk, $pkSize),
            'secretKey' => \FFI::string($sk, $skSize),
        ];
    }

    /**
     * Sign a message.
     *
     * @param string $message
     * @param string $secretKey
     * @return string The signature
     */
    public function sign(string $message, string $secretKey): string
    {
        $ffi = PQC::ffi();
        $maxSigSize = $this->maxSignatureSize();

        $sigBuf = $ffi->new("uint8_t[{$maxSigSize}]");
        $sigLen = $ffi->new("size_t");
        $sigLen->cdata = $maxSigSize;

        $msgLen = strlen($message);
        $msgBuf = $ffi->new("uint8_t[{$msgLen}]");
        \FFI::memcpy($msgBuf, $message, $msgLen);

        $skLen = strlen($secretKey);
        $skBuf = $ffi->new("uint8_t[{$skLen}]");
        \FFI::memcpy($skBuf, $secretKey, $skLen);

        $rc = $ffi->pqc_sig_sign($this->handle, $sigBuf, \FFI::addr($sigLen), $msgBuf, $msgLen, $skBuf);
        if ($rc !== 0) {
            throw new RuntimeException("Signature sign failed: " . $ffi->pqc_status_string($rc));
        }

        return \FFI::string($sigBuf, $sigLen->cdata);
    }

    /**
     * Verify a signature.
     *
     * @param string $message
     * @param string $signature
     * @param string $publicKey
     * @return bool True if valid
     */
    public function verify(string $message, string $signature, string $publicKey): bool
    {
        $ffi = PQC::ffi();

        $msgLen = strlen($message);
        $msgBuf = $ffi->new("uint8_t[{$msgLen}]");
        \FFI::memcpy($msgBuf, $message, $msgLen);

        $sigLen = strlen($signature);
        $sigBuf = $ffi->new("uint8_t[{$sigLen}]");
        \FFI::memcpy($sigBuf, $signature, $sigLen);

        $pkLen = strlen($publicKey);
        $pkBuf = $ffi->new("uint8_t[{$pkLen}]");
        \FFI::memcpy($pkBuf, $publicKey, $pkLen);

        $rc = $ffi->pqc_sig_verify($this->handle, $msgBuf, $msgLen, $sigBuf, $sigLen, $pkBuf);
        return $rc === 0;
    }
}

File: bindings/csharp/src/Signature.cs
// libpqc-dyber - Post-Quantum Cryptography Library
// Copyright (c) 2024-2026 Dyber, Inc.
// SPDX-License-Identifier: Apache-2.0 OR MIT

using System;
using System.Runtime.InteropServices;

namespace Dyber.PQC
{
    /// <summary>
    /// Digital signature operations.
    /// </summary>
    public sealed class Signature : IDisposable
    {
        private IntPtr _handle;
        private bool _disposed;

        /// <summary>
        /// Create a Signature context for the specified algorithm.
        /// </summary>
        /// <param name="algorithm">Algorithm name, e.g. "ML-DSA-65".</param>
        /// <exception cref="ArgumentException">Thrown when the algorithm is not supported.</exception>
        public Signature(string algorithm)
        {
            _handle = NativeMethods.pqc_sig_new(algorithm);
            if (_handle == IntPtr.Zero)
                throw new ArgumentException($"Unsupported or disabled signature algorithm: {algorithm}");
        }

        /// <summary>The algorithm name.</summary>
        public string AlgorithmName
        {
            get
            {
                ThrowIfDisposed();
                IntPtr ptr = NativeMethods.pqc_sig_algorithm(_handle);
                return Marshal.PtrToStringAnsi(ptr) ?? string.Empty;
            }
        }

        /// <summary>Public key size in bytes.</summary>
        public int PublicKeySize
        {
            get { ThrowIfDisposed(); return (int)(uint)NativeMethods.pqc_sig_public_key_size(_handle); }
        }

        /// <summary>Secret key size in bytes.</summary>
        public int SecretKeySize
        {
            get { ThrowIfDisposed(); return (int)(uint)NativeMethods.pqc_sig_secret_key_size(_handle); }
        }

        /// <summary>Maximum signature size in bytes.</summary>
        public int MaxSignatureSize
        {
            get { ThrowIfDisposed(); return (int)(uint)NativeMethods.pqc_sig_max_signature_size(_handle); }
        }

        /// <summary>NIST security level (1-5).</summary>
        public int SecurityLevel
        {
            get { ThrowIfDisposed(); return NativeMethods.pqc_sig_security_level(_handle); }
        }

        /// <summary>Whether this is a stateful signature algorithm.</summary>
        public bool IsStateful
        {
            get { ThrowIfDisposed(); return NativeMethods.pqc_sig_is_stateful(_handle) != 0; }
        }

        /// <summary>
        /// Generate a new keypair.
        /// </summary>
        /// <returns>A tuple of (publicKey, secretKey) byte arrays.</returns>
        public (byte[] PublicKey, byte[] SecretKey) Keygen()
        {
            ThrowIfDisposed();
            byte[] pk = new byte[PublicKeySize];
            byte[] sk = new byte[SecretKeySize];
            int rc = NativeMethods.pqc_sig_keygen(_handle, pk, sk);
            if (rc != 0)
                throw new InvalidOperationException($"Signature keygen failed: {GetStatusString(rc)}");
            return (pk, sk);
        }

        /// <summary>
        /// Sign a message.
        /// </summary>
        /// <param name="message">The message to sign.</param>
        /// <param name="secretKey">The signer's secret key.</param>
        /// <returns>The signature byte array (trimmed to actual length).</returns>
        public byte[] Sign(byte[] message, byte[] secretKey)
        {
            ThrowIfDisposed();
            if (message == null) throw new ArgumentNullException(nameof(message));
            if (secretKey == null) throw new ArgumentNullException(nameof(secretKey));
            if (secretKey.Length != SecretKeySize)
                throw new ArgumentException($"Secret key must be {SecretKeySize} bytes.");

            byte[] sig = new byte[MaxSignatureSize];
            UIntPtr sigLen = (UIntPtr)sig.Length;
            int rc = NativeMethods.pqc_sig_sign(_handle, sig, ref sigLen,
                message, (UIntPtr)message.Length, secretKey);
            if (rc != 0)
                throw new InvalidOperationException($"Signature sign failed: {GetStatusString(rc)}");

            int actualLen = (int)(uint)sigLen;
            if (actualLen < sig.Length)
            {
                byte[] trimmed = new byte[actualLen];
                Array.Copy(sig, trimmed, actualLen);
                return trimmed;
            }
            return sig;
        }

        /// <summary>
        /// Verify a signature.
        /// </summary>
        /// <param name="message">The message that was signed.</param>
        /// <param name="signature">The signature to verify.</param>
        /// <param name="publicKey">The signer's public key.</param>
        /// <returns>True if the signature is valid, false otherwise.</returns>
        public bool Verify(byte[] message, byte[] signature, byte[] publicKey)
        {
            ThrowIfDisposed();
            if (message == null) throw new ArgumentNullException(nameof(message));
            if (signature == null) throw new ArgumentNullException(nameof(signature));
            if (publicKey == null) throw new ArgumentNullException(nameof(publicKey));
            if (publicKey.Length != PublicKeySize)
                throw new ArgumentException($"Public key must be {PublicKeySize} bytes.");

            int rc = NativeMethods.pqc_sig_verify(_handle, message, (UIntPtr)message.Length,
                signature, (UIntPtr)signature.Length, publicKey);
            return rc == 0; // PQC_OK
        }

        private static string GetStatusString(int status)
        {
            IntPtr ptr = NativeMethods.pqc_status_string(status);
            return ptr != IntPtr.Zero ? Marshal.PtrToStringAnsi(ptr) ?? "unknown error" : "unknown error";
        }

        private void ThrowIfDisposed()
        {
            if (_disposed) throw new ObjectDisposedException(nameof(Signature));
        }

        public void Dispose()
        {
            if (!_disposed)
            {
                if (_handle != IntPtr.Zero)
                {
                    NativeMethods.pqc_sig_free(_handle);
                    _handle = IntPtr.Zero;
                }
                _disposed = true;
            }
        }

        ~Signature() => Dispose();
    }
}
